core/include/detray/grids/grid2.hpp

```diff
@@ -83,22 +90,21 @@ class grid2 {
 
     /** Constructor from grid data
      **/
-#if defined(__CUDACC__)
-    template <typename grid_view_t>
-    DETRAY_DEVICE grid2(grid_view_t &grid_data, const bare_value m_invalid =
-                                                    invalid_value<bare_value>())
+    template <typename grid_view_type>
+    DETRAY_DEVICE grid2(
+        grid_view_type &grid_data,
+        const bare_value m_invalid = invalid_value<bare_value>())
```

Comment:
I very much like that you removed the `#if defined(__CUDACC___)` from the code. (Since with the rewrite this is now possible.)

However you forgot to change `DETRAY_DEVICE` to `DETRAY_HOST_DEVICE`. :wink: I spent quite some time just today hunting down an inexplicable program termination in https://github.com/acts-project/algebra-plugins/pull/39 because of a similar mistake. Since as it turns out, when you mark a function only with `__device__`, CUDA will let you build your code without any complaints. But since the code of that function is not available for the host in the produced binary, the final executable exits without any error messages when you try to execute that function on the host. (Even `cuda-gdb` didn't say clearly what the issue was...)

Long story short, please change `DETRAY_DEVICE` to `DETRAY_HOST_DEVICE`. :wink:

Reply:
Hmm with `DETRAY_HOST_DEVICE`, all problems are solved?

> But since the code of that function is not available for the host in the produced binary, the final executable exits without any error messages when you try to execute that function on the host.

It's very weird. Given that the compiler is nvcc, shouldn't we get the error message of "it is defined only for the device"? It's a bug??